src/IO/H5/File.cpp

```diff
@@ -88,19 +88,26 @@ H5File<Access_t>& H5File<Access_t>::operator=(H5File&& rhs) {
 
 template <AccessType Access_t>
 H5File<Access_t>::~H5File() {
-  try {
-    close();
-  } catch (std::exception& ex) {
-    std::ostringstream os;
-    os << ex.what();
-    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
-    CkError("%s\n", os.str().c_str());
-  } catch (...) {
-    std::ostringstream os;
-    os << "Terminated due to unknown exception when closing H5 File "
-       << file_name_;
-    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
-    CkError("%s\n", os.str().c_str());
+  if (can_safely_throw_()) {
```

Comment:
But that's in a try block with a universal catch clause.  The exception can't escape.

Reply:
true, but the two branches differ in whether the code:
- terminates when the only error is failing to close the file, as I couldn't find a satisfactory way to throw the exception from the destructor
- prints a message that the file is being closed while another exception has been thrown, prints a message if the file is not closed successfully, but doesn't terminate.

So are you suggesting  I handle the conditional differently (i.e. not use CanSafelyThrow), or that I make some change to one of the branches?

Reply:
CanSafelyThrow is *only* for deciding whether to throw an exception from the destructor.  If you do not want to throw an exception from the destructor, do not use it.  If you're not throwing an exception, why do you want different behavior?

Reply:
I guess it depends upon how we want to handle the following situations:
1. With no previous exceptions thrown, the destructor fails to close the file
2. Another exception has been thrown, and while unwinding the stack the destructor fails to close the file

Possible choices:
   - In both cases, terminate with an error message about failing to close the file (This throws away the error message from another exception, and prevents attempting a controlled shutdown)
   - In both cases, print a diagnostic message about failing to close the file, but do not terminate. (Lets the code continue running with the file in an unknown state, but allow catching the exception)
   - Attempt to rethrow any exceptions before closing the file in the destructor, and use that to choose between terminating or printing a diagnostic message
   - Use CanSafelyThrow to (if it is safe) throw an exception from the destructor (instead of terminating) even though the destructor cannot be marked `noexcept(false)` (as that requires PUP::able to have its destructor marked `noexcept(false)`)

Do you have a preference for any of those choices, or an alternative suggestion?

Reply:
Abort unconditionally.  You can print `std::current_exception` if you want to.  If the hdf5 library isn't working you're not going to be able to do any meaningful cleanup anyway.